vcr/pe/presentation_submission.go

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package pe
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	v2 "github.com/nuts-foundation/nuts-node/vcr/pe/schema/v2"
+)
+
+// ParsePresentationSubmission validates the given JSON and parses it into a PresentationSubmission.
+// It returns an error if the JSON is invalid or doesn't match the JSON schema for a PresentationSubmission.
+func ParsePresentationSubmission(raw []byte) (*PresentationSubmission, error) {
+	enveloped := `{"presentation_submission":` + string(raw) + `}`
+	if err := v2.Validate([]byte(enveloped), v2.PresentationSubmission); err != nil {
+		return nil, err
+	}
+	var result PresentationSubmission
+	err := json.Unmarshal(raw, &result)
+	if err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+// PresentationSubmissionBuilder is a builder for PresentationSubmissions.
+// Multiple presentation definitions can be added to the builder.
+type PresentationSubmissionBuilder struct {
+	holders                []did.DID
+	presentationDefinition PresentationDefinition
+	wallets                [][]vc.VerifiableCredential
+}
+
+// Builder returns a new PresentationSubmissionBuilder.
+// A PresentationSubmissionBuilder can be used to create a PresentationSubmission with multiple wallets as input.
+func (presentationDefinition PresentationDefinition) Builder() PresentationSubmissionBuilder {
+	return PresentationSubmissionBuilder{
+		presentationDefinition: presentationDefinition,
+	}
+}
+
+// AddWallet adds credentials from a wallet that may be used to create the PresentationSubmission.
+func (b *PresentationSubmissionBuilder) AddWallet(holder did.DID, vcs []vc.VerifiableCredential) {
+	b.holders = append(b.holders, holder)
```

Comment:
note: we might have to support multiple holders later (match submission from 2 wallets of different holders).

Reply:
Which already works for the submission part. Only the VP creation part is missing, therefore the SignInstructions are added.